runtime/permissions.rs

```diff
@@ -63,515 +65,506 @@ impl Default for PermissionState {
       if state == PermissionState::Prompt {
-        if permission_prompt("Deno requests read access") {
-          self.read.granted_list.clear();
-          self.read.global_state = PermissionState::Granted;
-          return PermissionState::Granted;
+        if permission_prompt("write access") {
+          self.granted_list.clear();
+          self.global_state = PermissionState::Granted;
+          PermissionState::Granted
         } else {
-          self.read.global_state = PermissionState::Denied;
-          return PermissionState::Denied;
+          self.global_state = PermissionState::Denied;
+          PermissionState::Denied
         }
+      } else {
+        state
       }
-      state
     }
   }
 
-  pub fn request_write(&mut self, path: &Option<&Path>) -> PermissionState {
+  pub fn revoke(&mut self, path: &Option<&Path>) -> PermissionState {
     if let Some(path) = path {
-      let (resolved_path, display_path) = self.resolved_and_display_path(path);
-      let state = self.query_write(&Some(&resolved_path));
-      if state == PermissionState::Prompt {
-        if permission_prompt(&format!(
-          "Deno requests write access to \"{}\"",
-          display_path.display()
-        )) {
-          self
-            .write
-            .granted_list
-            .retain(|path| !path.starts_with(&resolved_path));
-          self.write.granted_list.insert(resolved_path);
-          return PermissionState::Granted;
-        } else {
-          self
-            .write
+      let path = resolve_from_cwd(path).unwrap();
+      self
+        .granted_list
+        .retain(|path_| !path_.0.starts_with(&path));
+    } else {
+      self.granted_list.clear();
+      if self.global_state == PermissionState::Granted {
+        self.global_state = PermissionState::Prompt;
+      }
+    }
+    self.query(path)
+  }
+
+  pub fn check(&self, path: &Path) -> Result<(), AnyError> {
+    let (resolved_path, display_path) = resolved_and_display_path(path);
+    self.query(&Some(&resolved_path)).check(
+      &format!("write to \"{}\"", display_path.display()),
+      self.name,
+    )
+  }
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Default, Deserialize)]
+pub struct NetPermission(pub String, pub Option<u16>);
+
+impl NetPermission {
+  fn new<T: AsRef<str>>(host: &&(T, Option<u16>)) -> Self {
+    NetPermission(host.0.as_ref().to_string(), host.1)
+  }
+
+  pub fn from_string(host: String) -> Self {
+    let url = url::Url::parse(&format!("http://{}", host)).unwrap();
+    let hostname = url.host_str().unwrap().to_string();
+
+    NetPermission(hostname, url.port())
+  }
+}
+
+impl fmt::Display for NetPermission {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    f.write_str(&match self.1 {
+      None => self.0.clone(),
+      Some(port) => format!("{}:{}", self.0, port),
+    })
+  }
+}
+
+impl UnaryPermission<NetPermission> {
+  pub fn query<T: AsRef<str>>(
+    &self,
+    host: &Option<&(T, Option<u16>)>,
```

Comment:
This type looks funky, why use reference to option and then reference to tuple?

Reply:
This is how it currently already is implemented. Not sure how simple it would be to change

Reply:
changed to `Option<&(T, Option<u16>)>`